util/Migrator/DbScripts/2020-03-26_00_CipherSoftDelete.sql

```diff
@@ -0,0 +1,483 @@
+    END
+
+    -- Cleanup orgs
+    DECLARE @OrgId UNIQUEIDENTIFIER
+    DECLARE [OrgCursor] CURSOR FORWARD_ONLY FOR
+        SELECT
+            [OrganizationId]
+        FROM
+            #Temp
+        WHERE
+            [OrganizationId] IS NOT NULL
+        GROUP BY
+            [OrganizationId]
+    OPEN [OrgCursor]
+    FETCH NEXT FROM [OrgCursor] INTO @OrgId
+    WHILE @@FETCH_STATUS = 0 BEGIN
+        -- Storage cleanup for groups only matters if we're permanently deleting
+        IF @Permanent = 1
+        BEGIN
+            EXEC [dbo].[Organization_UpdateStorage] @OrgId
+        END
+        EXEC [dbo].[User_BumpAccountRevisionDateByOrganizationId] @OrgId
+        FETCH NEXT FROM [OrgCursor] INTO @OrgId
+    END
+    CLOSE [OrgCursor]
+    DEALLOCATE [OrgCursor]
+
+    -- Cleanup user
+    IF @Permanent = 1
+    BEGIN
+        -- Storage cleanup for users only matters if we're permanently deleting
+        DECLARE @UserCiphersWithStorageCount INT
+        SELECT
+            @UserCiphersWithStorageCount = COUNT(1)
+        FROM
+            #Temp
+        WHERE
+            [UserId] IS NOT NULL
+            AND [Attachments] = 1
+
+        IF @UserCiphersWithStorageCount > 0
+        BEGIN
+            EXEC [dbo].[User_UpdateStorage] @UserId
+        END
+    END
+    EXEC [dbo].[User_BumpAccountRevisionDate] @UserId
+
+    DROP TABLE #Temp
+END
+GO
+
+IF OBJECT_ID('[dbo].[Cipher_DeleteById]') IS NOT NULL
+BEGIN
+    DROP PROCEDURE [dbo].[Cipher_DeleteById];
+END
+GO
+
+CREATE PROCEDURE [dbo].[Cipher_DeleteById]
+    @Id UNIQUEIDENTIFIER,
+    @Permanent AS BIT = 0
+WITH RECOMPILE
+AS
+BEGIN
+    SET NOCOUNT ON
+
+    DECLARE @UserId UNIQUEIDENTIFIER
+    DECLARE @OrganizationId UNIQUEIDENTIFIER
+    DECLARE @Attachments BIT
+
+    SELECT TOP 1
+        @UserId = [UserId],
+        @OrganizationId = [OrganizationId],
+        @Attachments = CASE WHEN [Attachments] IS NOT NULL THEN 1 ELSE 0 END
+    FROM
+        [dbo].[Cipher]
+    WHERE
+        [Id] = @Id
+        
+    IF @Permanent = 1
+    BEGIN
+        DELETE
+        FROM
+            [dbo].[Cipher]
+        WHERE
+            [Id] = @Id
+    END
+    ELSE
+    BEGIN
+        UPDATE
+            [dbo].[Cipher]
+        SET
+            [DeletedDate] = SYSUTCDATETIME()
+        WHERE
+            [Id] = @Id
+    END
+
+    IF @OrganizationId IS NOT NULL
+    BEGIN
+        IF @Attachments = 1 AND @Permanent = 1
+        BEGIN
+            EXEC [dbo].[Organization_UpdateStorage] @OrganizationId
+        END
+        EXEC [dbo].[User_BumpAccountRevisionDateByCipherId] @Id, @OrganizationId
+    END
+    ELSE IF @UserId IS NOT NULL
+    BEGIN
+        IF @Attachments = 1 AND @Permanent = 1
+        BEGIN
+            EXEC [dbo].[User_UpdateStorage] @UserId
+        END
+        EXEC [dbo].[User_BumpAccountRevisionDate] @UserId
+    END
+END
+GO
+
+IF OBJECT_ID('[dbo].[Cipher_Create]') IS NOT NULL
+BEGIN
+    EXECUTE sp_refreshsqlmodule N'[dbo].[Cipher_Create]';
+END
+GO
```

Comment:
Why did these need refreshing?

Reply:
When there are schema changes that impact a View, often times other stored procedures that reference the view need to have their query plans or updated to account for the new column/update. In this case SQL Server automatically detected these specific objects needed a module refresh when performing the schema diff, so I included them in the schema update accordingly.